Postgres with SSL throws "PSQLException: SSL error: null"

##### Bug Report

##### Summary

When using a Postgres database with SSL enabled, GoCD fails to start because it cannot establish a connection with the database.

##### Details
When GoCD server starts up, it seems to successfully connect to the database at first using SSL and run the db-migration checks:
```
ConnectionManager:110 - Loading database config from file config/db.properties
DatabaseMigrator:40 - Upgrading database, this might take a while depending on the size of the database.
************************************************************************ 
DatabaseMigrator:49 - ************************************************************************
WARNING: Shutting down your server at this point will lead to a database corruption. Please wait until the database upgrade completes.
DatabaseMigrator:49 - WARNING: Shutting down your server at this point will lead to a database corruption. Please wait until the database upgrade completes.
************************************************************************
DatabaseMigrator:49 - ************************************************************************
INFO: Database upgrade completed successfully.
DatabaseMigrator:57 - Database upgrade completed successfully.
DataMigrationRunner:34 - Running data migrations...
DataMigrationRunner:49 - Data migration took 3 ms
DataMigrationRunner:49 - Data migration took 2 ms
DataMigrationRunner:39 - Data migrations completed.
```
Then when GoCD tries to fetch pipelines, stages, etc., it fails with the exception:
```
...
Caused by: org.postgresql.util.PSQLException: SSL error: null 
     at org.postgresql.ssl.MakeSSL.convert(MakeSSL.java:43) 
...
```
Full stack trace is attached below.


###### Basic environment details

* Go Version: `20.8.0 (12213-1e23a06e496205ced5f1a8e83d9b209fc0a290cb)`
* JAVA Version: `14.0.2`
* OS: `Linux 4.19.112+`

##### Steps to Reproduce

Prerequisites: 
1. Set up postgres: https://docs.gocd.org/current/installation/configuring_database/postgres.html (Steps 1-2)
2. Obtain 3 files from Postgres DB (server certificate, client's public key, client's private key)
3. Convert them to DER format

Steps:
1. Modify config/db.properties so it would have all necessary settings to connect to Postgres with SSL enabled:
```
db.driver=org.postgresql.Driver
db.url=jdbc:postgresql://postgres.company.com:5432/gocd
db.user=gocd
db.password=password
db.connectionProperties.sslmode=verify-ca
db.connectionProperties.sslrootcert=/godata/config/db-certificates/server-public.der
db.connectionProperties.sslcert=/godata/config/db-certificates/gocd-public.der
db.connectionProperties.sslkey=/godata/config/db-certificates/devops-postgres-db-gocd-private.pk8

```
2. Restart GoCD server
3. Observe multiple exception during startup in the logs.


##### Actual Results
Exceptions during startup:
```
Caused by: java.lang.RuntimeException: org.springframework.dao.DataAccessResourceFailureException: could not execute update query; nested exception is org.hibernate.exception.JDBCConnectionException: could not execute update query 
	at com.thoughtworks.go.server.initializers.ApplicationInitializer.onApplicationEvent(ApplicationInitializer.java:158) 
	at com.thoughtworks.go.server.initializers.ApplicationInitializer.onApplicationEvent(ApplicationInitializer.java:49) 
	at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:172) 
	at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:165) 
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:139) 
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:393) 
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:347) 
	at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:883) 
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:545) 
	at org.springframework.web.context.ContextLoader.configureAndRefreshWebApplicationContext(ContextLoader.java:443) 
	at org.springframework.web.context.ContextLoader.initWebApplicationContext(ContextLoader.java:325) 
	at org.springframework.web.context.ContextLoaderListener.contextInitialized(ContextLoaderListener.java:107) 
	at org.eclipse.jetty.server.handler.ContextHandler.callContextInitialized(ContextHandler.java:933) 
	at org.eclipse.jetty.servlet.ServletContextHandler.callContextInitialized(ServletContextHandler.java:553) 
	at org.eclipse.jetty.server.handler.ContextHandler.startContext(ContextHandler.java:892) 
	at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:356) 
	at org.eclipse.jetty.webapp.WebAppContext.startWebapp(WebAppContext.java:1445) 
	at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1409) 
	at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:825) 
	at org.eclipse.jetty.servlet.ServletContextHandler.doStart(ServletContextHandler.java:275) 
	at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:524) 
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:72) 
	at org.eclipse.jetty.deploy.bindings.StandardStarter.processBinding(StandardStarter.java:46) 
	at org.eclipse.jetty.deploy.AppLifeCycle.runBindings(AppLifeCycle.java:188) 
	at org.eclipse.jetty.deploy.DeploymentManager.requestAppGoal(DeploymentManager.java:513) 
	at org.eclipse.jetty.deploy.DeploymentManager.addApp(DeploymentManager.java:154) 
	at com.thoughtworks.go.server.Jetty9Server.startHandlers(Jetty9Server.java:179) 
	at com.thoughtworks.go.server.Jetty9Server.start(Jetty9Server.java:129) 
	at com.thoughtworks.go.server.GoServer.startServer(GoServer.java:62) 
	... 14 more 
Caused by: org.springframework.dao.DataAccessResourceFailureException: could not execute update query; nested exception is org.hibernate.exception.JDBCConnectionException: could not execute update query 
	at org.springframework.orm.hibernate3.SessionFactoryUtils.convertHibernateAccessException(SessionFactoryUtils.java:639) 
	at org.springframework.orm.hibernate3.HibernateAccessor.convertHibernateAccessException(HibernateAccessor.java:416) 
	at org.springframework.orm.hibernate3.HibernateTemplate.doExecute(HibernateTemplate.java:416) 
	at org.springframework.orm.hibernate3.HibernateTemplate.execute(HibernateTemplate.java:342) 
	at com.thoughtworks.go.server.persistence.ServerBackupRepository.markInProgressBackupsAsAborted(ServerBackupRepository.java:65) 
	at com.thoughtworks.go.server.service.BackupService.initialize(BackupService.java:117) 
	at com.thoughtworks.go.server.initializers.ApplicationInitializer.onApplicationEvent(ApplicationInitializer.java:154) 
	... 42 more 
Caused by: org.hibernate.exception.JDBCConnectionException: could not execute update query 
	at org.hibernate.exception.SQLStateConverter.convert(SQLStateConverter.java:99) 
	at org.hibernate.exception.JDBCExceptionHelper.convert(JDBCExceptionHelper.java:66) 
	at org.hibernate.hql.ast.exec.BasicExecutor.execute(BasicExecutor.java:110) 
	at org.hibernate.hql.ast.QueryTranslatorImpl.executeUpdate(QueryTranslatorImpl.java:423) 
	at org.hibernate.engine.query.HQLQueryPlan.performExecuteUpdate(HQLQueryPlan.java:283) 
	at org.hibernate.impl.SessionImpl.executeUpdate(SessionImpl.java:1288) 
	at org.hibernate.impl.QueryImpl.executeUpdate(QueryImpl.java:117) 
	at com.thoughtworks.go.server.persistence.ServerBackupRepository.lambda$markInProgressBackupsAsAborted$2(ServerBackupRepository.java:70) 
	at org.springframework.orm.hibernate3.HibernateTemplate.doExecute(HibernateTemplate.java:411) 
	... 46 more 
Caused by: org.postgresql.util.PSQLException: SSL error: null 
	at org.postgresql.ssl.MakeSSL.convert(MakeSSL.java:43) 
	at org.postgresql.core.v3.ConnectionFactoryImpl.enableSSL(ConnectionFactoryImpl.java:445) 
	at org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:139) 
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:196) 
	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49) 
	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:211) 
	at org.postgresql.Driver.makeConnection(Driver.java:459) 
	at org.postgresql.Driver.connect(Driver.java:261) 
	at org.apache.commons.dbcp2.DriverConnectionFactory.createConnection(DriverConnectionFactory.java:55) 
	at org.apache.commons.dbcp2.PoolableConnectionFactory.makeObject(PoolableConnectionFactory.java:355) 
	at org.apache.commons.pool2.impl.GenericObjectPool.create(GenericObjectPool.java:899) 
	at org.apache.commons.pool2.impl.GenericObjectPool.borrowObject(GenericObjectPool.java:429) 
	at org.apache.commons.pool2.impl.GenericObjectPool.borrowObject(GenericObjectPool.java:354) 
	at org.apache.commons.dbcp2.PoolingDataSource.getConnection(PoolingDataSource.java:134) 
	at org.apache.commons.dbcp2.BasicDataSource.getConnection(BasicDataSource.java:753) 
	at org.springframework.jdbc.datasource.DataSourceUtils.doGetConnection(DataSourceUtils.java:111) 
	at org.springframework.jdbc.datasource.TransactionAwareDataSourceProxy$TransactionAwareInvocationHandler.invoke(TransactionAwareDataSourceProxy.java:229) 
	at com.sun.proxy.$Proxy34.prepareStatement(Unknown Source) 
	at org.hibernate.jdbc.AbstractBatcher.getPreparedStatement(AbstractBatcher.java:534) 
	at org.hibernate.jdbc.AbstractBatcher.prepareStatement(AbstractBatcher.java:116) 
	at org.hibernate.jdbc.AbstractBatcher.prepareStatement(AbstractBatcher.java:109) 
	at org.hibernate.hql.ast.exec.BasicExecutor.execute(BasicExecutor.java:88) 
	... 52 more 
Caused by: javax.net.ssl.SSLException 
	at java.base/sun.security.ssl.Alert.createSSLException(Unknown Source) 
	at java.base/sun.security.ssl.TransportContext.fatal(Unknown Source) 
	at java.base/sun.security.ssl.TransportContext.fatal(Unknown Source) 
	at java.base/sun.security.ssl.TransportContext.fatal(Unknown Source) 
	at java.base/sun.security.ssl.SSLSocketImpl.handleException(Unknown Source) 
	at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(Unknown Source) 
	at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(Unknown Source) 
	at org.postgresql.ssl.MakeSSL.convert(MakeSSL.java:41) 
	... 73 more 
Caused by: java.lang.NullPointerException 
	at org.bouncycastle.crypto.signers.PSSSigner.generateSignature(Unknown Source) 
	at org.bouncycastle.jcajce.provider.asymmetric.rsa.PSSSignatureSpi.engineSign(Unknown Source) 
	at java.base/java.security.Signature$Delegate.engineSign(Unknown Source) 
	at java.base/java.security.Signature.sign(Unknown Source) 
	at java.base/sun.security.ssl.CertificateVerify$T13CertificateVerifyMessage.<init>(Unknown Source) 
	at java.base/sun.security.ssl.CertificateVerify$T13CertificateVerifyProducer.onProduceCertificateVerify(Unknown Source) 
	at java.base/sun.security.ssl.CertificateVerify$T13CertificateVerifyProducer.produce(Unknown Source) 
	at java.base/sun.security.ssl.SSLHandshake.produce(Unknown Source) 
	at java.base/sun.security.ssl.Finished$T13FinishedConsumer.onConsumeFinished(Unknown Source) 
	at java.base/sun.security.ssl.Finished$T13FinishedConsumer.consume(Unknown Source) 
	at java.base/sun.security.ssl.SSLHandshake.consume(Unknown Source) 
	at java.base/sun.security.ssl.HandshakeContext.dispatch(Unknown Source) 
	at java.base/sun.security.ssl.HandshakeContext.dispatch(Unknown Source) 
	at java.base/sun.security.ssl.TransportContext.dispatch(Unknown Source) 
	at java.base/sun.security.ssl.SSLTransport.decode(Unknown Source) 
	at java.base/sun.security.ssl.SSLSocketImpl.decode(Unknown Source) 
	at java.base/sun.security.ssl.SSLSocketImpl.readHandshakeRecord(Unknown Source) 
	... 76 more 
```

##### Possible Fix
It seems to be a known bug in bouncy castle library and it has been already fixed in release 1.65: https://github.com/bcgit/bc-java/issues/633
Updating the dependency to the latest version of bouncycastle should fix the issue.

--- a/build.gradle
+++ b/build.gradle
@@ -170,7 +170,7 @@ rootProject.ext.versions = [
   aspectj               : '1.9.5',
   assertJ               : '3.12.2',
   assertJ_DB            : '1.3.0',
-  bouncyCastle          : '1.59', // bouncycastle version has to be compatible with the jruby version
+  bouncyCastle          : '1.65', // bouncycastle version has to be compatible with the jruby version
   bundler               : '2.1.1',
   cglib                 : '3.3.0',
   cloning               : '1.9.11',

--- a/server/src/main/webapp/WEB-INF/rails/Gemfile
+++ b/server/src/main/webapp/WEB-INF/rails/Gemfile
@@ -15,6 +15,7 @@ gem 'representable', '2.1.7'
 gem 'js-routes'
 gem 'ts_routes'
 gem 'versionist'
+gem 'jruby-openssl', '0.10.5' #Check BC version before upgrading, require '0.10.5' since it's compatible with Bouncy Castle 1.65
 
 # Windows does not include zoneinfo files, so bundle the tzinfo-data gem
 gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]

